### ISSUE TITLE
compiling older commits for ASV needs old cython version

### ISSUE THREAD

**Issue by orbeckst** [MDAnalysis/benchmarks#7]
I wanted to fill in some gaps in the benchmarking history but it seems that with my current ASV setup I am not able anymore to build MDAnalysis at some earlier snapshots. For example, to fill in a gap (visible in [analysis.psa.PSA_metricBench.time_hausdorff near release 0.17.0](https://www.mdanalysis.org/benchmarks/#analysis.psa.PSA_metricBench.time_hausdorff)) I wanted to benchmark commit MDAnalysis/mdanalysis@b56ef0165d814f898590eb18c801ec5a7b385f21. ASV fails and skips everything because MDAnalysis fails to build. From cursory inspection this is a problem with cythonizing `encore/clustering/affinityprop.pyx`. This seems related to the cython 0.27/0.28 snafu (https://github.com/MDAnalysis/mdanalysis/pull/1833)  – below is the output from trying to build locally. Building current develop works fine.

Building the old commit with cython pinned to 0.27 also works.

Is there a way to tell ASV to use old cython until we hit MDAnalysis/mdanalysis@74defb7fb168145e58388cc67af811559cf23921  when we switched over to cython 0.28?

If we don’t solve this problem we will not be able to benchmark older commits with newly added benchmarks.

(Posted to the developer list on July 9, see https://groups.google.com/d/msg/mdnalysis-devel/divONaOi2Wc/3uKo2KqsAwAJ)

**Comment by orbeckst** [comment]
Well... this problem  just wiped out a good part of the published benchmark history 65ef2cadb990088e069cc11242ce47875f69a22d. (Older tests re-ran #8 , but then couldn't build the project and flagged the whole commit as failed.)

I can restore an earlier commit but this problem needs to be fixed ASAP.

**Comment by orbeckst** [comment]
ping @tylerjereddy – insights welcome!

Is there an ASV mailing list??

**Comment by tylerjereddy** [comment]
@orbeckst The idea of providing an `asv` API for variable dependency usage on some kind of per-commit basis was [discussed fairly recently](https://github.com/scipy/scipy-articles/pull/15#issuecomment-377445436) in the prep of the SciPy manuscript--I had to do many different runs with different versions of Cython pinned there to adequately sample history due to build failures with Cython versions.

I can ping @pv for you, but he is beyond spread thin -- I just open issues on `asv` tracker and Pauli tags them as questions; that's perhaps ok as long as it isn't done too often. Not a substitute for reading the docs, obviously, because it is basically just him answering.

In short, I don't think there's a built-in way to do this within `asv` at the moment (and the discussion thread above reflects that I think).

Remind me--does it not work to put a list of Cython versions in the dependency matrix? I know this is supported according to docs. Are you willing to pay the computational cost of duplicating the benchmark matrix across all commits to cover this scenario? (or at least until asv supports something better?).

**Comment by orbeckst** [comment]
I thought about just increasing the matrix – the cost is not a problem at the moment.

I'll try it next.

**Comment by orbeckst** [comment]
I reverted the worst commit and stopped the automated benchmarks until I get multiple cythons benchmarked.

**Comment by orbeckst** [comment]
I started a complete rerun with cython 0.27 and the latest. It has to run  51680 total benchmarks (340 commits * 2 environments * 76 benchmarks), so it will probably take 1-2 days.

**Comment by tylerjereddy** [comment]
Hopefully this will eventually be something that can be controlled via `asv`; there are plans to do an `asv` sprint sometime in the near future. There seems to be some interest in having the ability to somehow benchmark each commit with its dependencies at i.e., release time / close to it (time-encapsulated ecosystem-based performance I suppose, which has the advantage that it should build since that's the release set of deps).

The other obvious issue with current approach is that even if it does work somebody will have to handle the maintenance burden of matrix expansion over time.

**Comment by orbeckst** [comment]
I was thinking about this, I would try to do the matrix with additional commit ranges
```json
"matrix": {
   "Cython": [ 
        {"release-0.11.0..beeeeef": "0.27", 
          "beeeef....deadbee": "0.28", 
          "deadbee..HEAD": ""},
        {"": "0.21"}
    ]
```

**Comment by tylerjereddy** [comment]
Ah yeah, that's probably more tractable than trying to automagically infer appropriate deps from  project data / metadata.

**Comment by orbeckst** [comment]
Using a matrix with cython/latest and cython/0.27 produces results. We have a full history again but at the price of a separation in cython `[default]` (i.e., latest > 0.27) and cython `[0.27]`, see eg https://www.mdanalysis.org/benchmarks/#/summarylist .

Somewhat surprisingly, all benchmarks seem to be running under cython 0.27 fine, at least for now.

I enable the cron job again.

Until a better solution comes along I will consider this issue "band-aid" fixed.
